src/services/factories/tokenList.ts

```diff
@@ -88,32 +99,124 @@ export function getTokensFactory(factoryParams: {
     // if something failed...
     if (failedToUpdate.length > 0) {
       logDebug(
-        `[network:${networkId}] Failed to fetch ids for ${failedToUpdate.length} tokens. Trying again in ${retryDelay}ms`,
+        `[tokenListFactory][${networkId}] Failed to fetch ids for ${failedToUpdate.length} tokens. Trying again in ${retryDelay}ms`,
       )
       // calculate how long to wait next time
       const nextDelay = retryDelay * (exponentialBackOff ? 2 : 1)
       // try again failed tokens after retryDelay
-      setTimeout(() => updateWithRetries(networkId, failedToUpdate, maxRetries - 1, nextDelay), retryDelay)
+      setTimeout(() => updateTokenIds(networkId, failedToUpdate, maxRetries - 1, nextDelay), retryDelay)
     }
   }
 
-  async function updateTokenIds(networkId: number): Promise<void> {
-    // Set as updated on start to prevent concurrent queries
-    updatedIdsForNetwork.add(networkId)
+  async function getErc20DetailsOrAddress(networkId: number, tokenAddress: string): Promise<TokenFromErc20 | string> {
+    // Simple wrapper function to return original address instead of null for make logging easier
+    const erc20Details = await getTokenFromErc20({ networkId, tokenAddress })
+    return erc20Details || tokenAddress
+  }
+
+  async function fetchAddressesAndIds(networkId: number, numTokens: number): Promise<Map<string, number>> {
+    logDebug(`[tokenListFactory][${networkId}] Fetching addresses for ids from 0 to ${numTokens - 1}`)
+
+    const promises = Array.from({ length: numTokens }, (_, tokenId) =>
+      exchangeApi.getTokenAddressById({ networkId, tokenId }),
+    )
+
+    const tokenAddresses = await Promise.all(promises)
+
+    return new Map(tokenAddresses.map((tokenAddress, id) => [tokenAddress, id]))
+  }
+
+  async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
+    // Fetch addresses from contract given numTokens count
+    const addressesAndIds = await retry({
+      fn: fetchAddressesAndIds,
+      fnParams: [networkId, numTokens],
+    })
```

Comment:
I would prefer a simpler kind of signature:
```ts
const addressesAndIds = await retry(() => fetchAddressesAndIds(networkId, numTokens))
```
Easier to reason about, imho

Reply:
agree

Reply:
The idea was to have the nice fn name when logging https://github.com/gnosis/dex-react/pull/967/files/9eddabf6906da351a54211bb43f8a8f498d020d1#diff-ee40819644f3b6e21adb3f10e5481ff4R134

But I see the point, will simplify.

Reply:
Aligned with my proposal above too

Reply:
Wait, maybe we don't say exactly the same. I was suggesting a HoF approach, so we can wrap functions? does it make sense? something like the one you've made for the price feed